src/ripple/app/misc/impl/ValidatorSite.cpp

```diff
@@ -165,153 +224,224 @@ ValidatorSite::onTimer (
+    }
+    else if (ListDisposition::invalid == disp)
+    {
+        JLOG (j_.warn()) <<
+            "Invalid validator list from " <<
+            sites_[siteIdx].activeResource->uri;
+    }
+    else if (ListDisposition::unsupported_version == disp)
+    {
+        JLOG (j_.warn()) <<
+            "Unsupported version validator list from " <<
+            sites_[siteIdx].activeResource->uri;
     }
     else
     {
-        sp = std::make_shared<detail::WorkPlain>(
-            sites_[siteIdx].pUrl.domain,
-            sites_[siteIdx].pUrl.path,
-            std::to_string(*sites_[siteIdx].pUrl.port),
-            ios_,
-            [this, siteIdx](error_code const& err, detail::response_type&& resp)
-            {
-                onSiteFetch (err, std::move(resp), siteIdx);
-            });
+        BOOST_ASSERT(false);
     }
 
-    work_ = sp;
-    sp->run ();
+    if (body.isMember ("refresh_interval") &&
+        body["refresh_interval"].isNumeric ())
+    {
+        // TODO: should we sanity check/clamp this value
+        // to something reasonable?
+        sites_[siteIdx].refreshInterval =
+            std::chrono::minutes{body["refresh_interval"].asUInt ()};
+    }
 }
 
-void
-ValidatorSite::onSiteFetch(
-    boost::system::error_code const& ec,
-    detail::response_type&& res,
-    std::size_t siteIdx)
+ValidatorSite::Site::ResourcePtr
+ValidatorSite::processRedirect (
+    detail::response_type& res,
+    std::size_t siteIdx,
+    std::lock_guard<std::mutex>& lock)
 {
-    if (! ec && res.result() != boost::beast::http::status::ok)
+    using namespace boost::beast::http;
+    Site::ResourcePtr newLocation;
+    if (res.find(field::location) == res.end() ||
+        res[field::location].empty())
     {
-        std::lock_guard <std::mutex> lock{sites_mutex_};
         JLOG (j_.warn()) <<
             "Request for validator list at " <<
-            sites_[siteIdx].uri << " returned " << res.result_int();
-
-        sites_[siteIdx].lastRefreshStatus.emplace(
-                Site::Status{clock_type::now(), ListDisposition::invalid});
+            sites_[siteIdx].activeResource->uri <<
+            " returned a redirect with no Location.";
+        throw std::runtime_error{"missing location"};
     }
-    else if (! ec)
+
+    if (sites_[siteIdx].redirCount == MAX_REDIRECTS)
     {
-        std::lock_guard <std::mutex> lock{sites_mutex_};
-        Json::Reader r;
-        Json::Value body;
-        if (r.parse(res.body().data(), body) &&
-            body.isObject () &&
-            body.isMember("blob") && body["blob"].isString () &&
-            body.isMember("manifest") && body["manifest"].isString () &&
-            body.isMember("signature") && body["signature"].isString() &&
-            body.isMember("version") && body["version"].isInt())
-        {
+        JLOG (j_.warn()) <<
+            "Exceeded max redirects for validator list at " <<
+            sites_[siteIdx].loadedResource->uri ;
```

Comment:
Should this be `sites_[siteIdx].activeResource->uri` ?

Reply:
it's a redirect loop of some kind, so I want to log the original URI (from the config) so that someone can investigate based on the starting point. `activeResource` will be the last uri we tried before giving up.